src/main/java/com/powsybl/openloadflow/network/impl/LfNetworkLoaderImpl.java

```diff
@@ -39,21 +43,82 @@
         private final Set<ThreeWindingsTransformer> t3wtSet = new LinkedHashSet<>();
     }
 
-    private static void createBuses(List<Bus> buses, LfNetworkParameters parameters, LfNetwork lfNetwork,
+    private static void createBuses(List<Bus> buses, LfNetworkParameters parameters, LfNetwork lfNetwork, List<LfBus> lfBuses,
                                     LoadingContext loadingContext, LfNetworkLoadingReport report) {
-        Map<LfBusImpl, String> controllerBusToControlledBusId = new LinkedHashMap<>();
-
         for (Bus bus : buses) {
-            LfBusImpl lfBus = createBus(bus, parameters, lfNetwork, loadingContext, report, controllerBusToControlledBusId);
+            LfBusImpl lfBus = createBus(bus, parameters, lfNetwork, loadingContext, report);
             lfNetwork.addBus(lfBus);
+            lfBuses.add(lfBus);
         }
+    }
 
+    private static void createVoltageControls(LfNetwork lfNetwork, List<LfBus> lfBuses, boolean voltageRemoteControl) {
         // set controller -> controlled link
-        for (Map.Entry<LfBusImpl, String> e : controllerBusToControlledBusId.entrySet()) {
-            LfBusImpl controllerBus = e.getKey();
-            String controlledBusId = e.getValue();
-            LfBus controlledBus = lfNetwork.getBusById(controlledBusId);
-            controllerBus.setControlledBus((LfBusImpl) controlledBus);
+        for (LfBus controllerBus : lfBuses) {
+
+            List<LfGenerator> voltageControlGenerators = controllerBus.getGenerators().stream().filter(LfGenerator::hasVoltageControl).collect(Collectors.toList());
+            if (!voltageControlGenerators.isEmpty()) {
+
+                LfGenerator lfGenerator0 = voltageControlGenerators.get(0);
+                LfBus controlledBus = lfGenerator0.getControlledBus(lfNetwork);
+                double controllerTargetV = lfGenerator0.getTargetV();
+
+                LfBus finalControlledBus = controlledBus;
+                voltageControlGenerators.stream().skip(1).forEach(lfGenerator -> {
+                    LfBus generatorControlledBus = lfGenerator.getControlledBus(lfNetwork);
+
+                    // check that remote control bus is the same for the generators of current controller bus which have voltage control on
+                    checkUniqueControlledBus(finalControlledBus, generatorControlledBus, controllerBus);
+
+                    // check that target voltage is the same for the generators of current controller bus which have voltage control on
+                    checkUniqueTargetVControllerBus(lfGenerator, controllerTargetV, controllerBus, generatorControlledBus);
+                });
+
+                if (!voltageRemoteControl && controlledBus != controllerBus) {
+                    // if voltage remote control deactivated and remote control, set local control instead
```

Comment:
This could be done earlier to have the lfGenerator consistent with the local voltage control, but for now it's easier to do it here.